pkg/virtualmachine/enricher/enricher_impl.go

```diff
@@ -2,52 +2,119 @@ package enricher
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/pkg/errors"
 	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errorhelpers"
 	"github.com/stackrox/rox/pkg/logging"
+	pkgScanners "github.com/stackrox/rox/pkg/scanners"
 	"github.com/stackrox/rox/pkg/scanners/types"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
 var (
 	log = logging.LoggerForModule()
 )
 
 type enricherImpl struct {
-	vmScanner types.VirtualMachineScanner
+	scanners map[string]types.VirtualMachineScanner
+	creators map[string]pkgScanners.VirtualMachineScannerCreator
+
+	lock sync.RWMutex
 }
 
-func New(scanner types.VirtualMachineScanner) VirtualMachineEnricher {
-	return &enricherImpl{
-		vmScanner: scanner,
+func newWithCreator(fn func() (string, func(*storage.ImageIntegration) (types.VirtualMachineScanner, error))) VirtualMachineEnricher {
+	enricher := &enricherImpl{
+		scanners: make(map[string]types.VirtualMachineScanner),
+		creators: make(map[string]pkgScanners.VirtualMachineScannerCreator),
 	}
+	name, creator := fn()
+	enricher.creators[name] = creator
+	return enricher
 }
 
 func (e *enricherImpl) EnrichVirtualMachineWithVulnerabilities(vm *storage.VirtualMachine, indexReport *v4.IndexReport) error {
 	// Clear any pre-existing notes
 	vm.Notes = vm.GetNotes()[:0]
 
-	if e.vmScanner == nil {
+	scanners := e.getScanners()
+	if len(scanners) == 0 {
 		vm.Notes = append(vm.Notes, storage.VirtualMachine_MISSING_SCAN_DATA)
-		return errors.New("Scanner V4 client not available for VM enrichment")
+		return errors.New("no scanner V4 client available for virtual machine enrichment")
+	}
+
+	errorList := errorhelpers.NewErrorList(fmt.Sprintf("error scanning virtual machine %s:%s", vm.GetClusterName(), vm.GetName()))
+	for _, scanner := range scanners {
+		if err := enrichVirtualMachineWithScanner(vm, indexReport, scanner); err != nil {
+			errorList.AddError(err)
```

Comment:
Let's make sure that we know which scanner is returning the error (if we have more than 1).